core/clustersmngr/factory.go

```diff
@@ -469,6 +494,102 @@ func (cf *clustersManager) userNsList(ctx context.Context, user *auth.UserPrinci
 	return cf.GetUserNamespaces(user)
 }
 
+func (cf *clustersManager) getOrCreateClient(ctx context.Context, user *auth.UserPrincipal, cfgFunc ClusterClientConfigFunc, cluster Cluster) (client.Client, error) {
+	if user == nil {
+		user = &auth.UserPrincipal{
+			ID: "weave-gitops-server",
+		}
+	}
+
+	if client, found := cf.usersClients.Get(user, cluster.Name); found {
+		return client, nil
+	}
+
+	client, err := cf.createClient(user, cfgFunc, cluster, cf.scheme)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating client for cluster=%s: %w", cluster.Name, err)
+	}
+
+	cf.usersClients.Set(user, cluster.Name, client)
+
+	return client, nil
+}
+
+func ClientFactory(user *auth.UserPrincipal, cfgFunc ClusterClientConfigFunc, cluster Cluster, scheme *apiruntime.Scheme) (client.Client, error) {
+	config, err := cfgFunc(cluster)
+	if err != nil {
+		return nil, fmt.Errorf("error building cluster client config: %w", err)
+	}
+
+	mapper, err := apiutil.NewDiscoveryRESTMapper(config)
+	if err != nil {
+		return nil, fmt.Errorf("could not create RESTMapper from config: %w", err)
+	}
+
+	client, err := client.New(config, client.Options{
+		Scheme: scheme,
+		Mapper: mapper,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create leaf client: %w", err)
+	}
+
+	return client, nil
```

Comment:
this code is common between Client and  CachedClient, do you think might be useful to refactor it to a shared function?